dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/ServiceDiscoveryRegistryDirectory.java

```diff
@@ -246,12 +246,17 @@ private void refreshInvoker(List<URL> invokerUrls) {
             destroyAllInvokers(); // Close all invokers
         } else {
             this.forbidden = false; // Allow accessing
-            Map<String, Invoker<T>> oldUrlInvokerMap = this.urlInvokerMap; // local reference
             if (CollectionUtils.isEmpty(invokerUrls)) {
                 return;
             }
 
-            Map<String, Invoker<T>> newUrlInvokerMap = toInvokers(invokerUrls);// Translate url list to Invoker map
+            // can't use local reference because this.urlInvokerMap might be accessed at isAvailable() by main thread concurrently.
+            Map<String, Invoker<T>> oldUrlInvokerMap = null;
+            if (this.urlInvokerMap != null) {
+                oldUrlInvokerMap = new HashMap<>();
```

Comment:
it would be better to sepcify the default size for HashMap here

Reply:
using LinkedHashMap will reduce temporally memery usage when using iterator in forEach

Reply:
@AlbumenJ already changed to LinkedHashMap and set the default size